vm/src/hint_processor/builtin_hint_processor/bootloader/program_loader.rs

```diff
@@ -0,0 +1,332 @@
+use crate::hint_processor::builtin_hint_processor::bootloader::types::BootloaderVersion;
+use crate::serde::deserialize_program::BuiltinName;
+use crate::types::program::Program;
+use crate::types::relocatable::Relocatable;
+use crate::vm::errors::hint_errors::HintError;
+use crate::vm::errors::memory_errors::MemoryError;
+use crate::vm::vm_memory::memory::Memory;
+use felt::Felt252;
+
+#[derive(thiserror_no_std::Error, Debug)]
+pub enum ProgramLoaderError {
+    #[error(transparent)]
+    Memory(#[from] MemoryError),
+
+    #[error("Program has no entrypoint")]
+    NoEntrypoint,
+}
+
+impl Into<HintError> for ProgramLoaderError {
+    fn into(self) -> HintError {
+        match self {
+            ProgramLoaderError::Memory(e) => HintError::Memory(e),
+            _ => HintError::CustomHint(self.to_string().into_boxed_str()),
+        }
+    }
+}
+
+/// Creates an instance of `Felt252` from a builtin name.
+///
+/// Converts the builtin name to bytes then attempts to create a felt from
+/// these bytes. This function will fail if the builtin name is over 31 characters.
+fn builtin_to_felt(builtin: &BuiltinName) -> Result<Felt252, ProgramLoaderError> {
+    // The Python implementation uses the builtin name without suffix
+    let builtin_name = builtin
+        .name()
+        .strip_suffix("_builtin")
+        .unwrap_or(builtin.name());
+
+    Ok(Felt252::from_bytes_be(builtin_name.as_bytes()))
+}
+
+pub struct LoadedProgram {
+    /// Start of the program code in the VM memory.
+    pub code_address: Relocatable,
+    /// Total size of the program in memory, header included.
+    pub size: usize,
+}
+
+/// Loads a Cairo program in the VM memory.
+pub struct ProgramLoader<'vm> {
+    /// Memory accessor.
+    memory: &'vm mut Memory,
```

Comment:
Would it simplify things to store a `Vm` instead?

Reply:
We only have access to an `&mut VirtualMachine` anyway, so we'd still have to play with references. We only need a `&mut Memory` so I think this is the right level of abstraction. 

I hesitated to pass `&memory` as a parameter to `load_program()` instead to only need an immutable reference to `ProgramLoader`, but I decided that it was alright as the VM object is doing the same thing anyway.